ruby_event_store/lib/ruby_event_store/specification.rb

```diff
@@ -190,10 +245,49 @@ def first
     #
     # @return [Event, nil]
     def last
-      record = repository.read(read_last.result)
+      record = repository.read(read_last)
       mapper.serialized_record_to_event(record) if record
     end
 
+    # Two read specifications are equal if:
+    # * they are of the same class
+    # * have identical data (verified with eql? method)
+    #
+    # @param other_spec [Specification, Object] object to compare
+    #
+    # @return [TrueClass, FalseClass]
+    def ==(other_spec)
+      other_spec.hash.eql?(hash)
+    end
+
+    # @private
+    BIG_VALUE = 0b100010010100011110111101100001011111100101001010111110101000000
+
+    # Generates a Fixnum hash value for this object. This function
+    # have the property that a.eql?(b) implies a.hash == b.hash.
+    #
+    # The hash value is used along with eql? by the Hash class to
+    # determine if two objects reference the same hash key.
+    #
+    # This hash is based on
+    # * class
+    # * stream_name
+    # * start
+    # * direction
+    # * read_as
+    # * batch_size
+    def hash
```

Comment:
Could we use all `instance_variables` ? Would that be safer? It's hard for me to say if we forgot to include something or not.